modules/git/blob_nogogit.go

```diff
@@ -54,11 +55,14 @@ func (b *Blob) DataAsync() (io.ReadCloser, error) {
 		return io.NopCloser(bytes.NewReader(bs)), err
 	}
 
-	return &blobReader{
+	br := &blobReader{
 		rd:     rd,
 		n:      size,
 		cancel: cancel,
-	}, nil
+	}
+	runtime.SetFinalizer(br, func(br *blobReader) { br.Close() })
```

Comment:
Hmm, these `SetFinalizer` have been extracted into a new PR 😊

Reply:
It will only be released when application exit. I don't think it's an expected behaviour?

Reply:
No - the SetFinalizer will run when the object is gc'd:

> SetFinalizer sets the finalizer associated with obj to the provided finalizer function. When the garbage collector finds an unreachable block with an associated finalizer, it clears the association and runs finalizer(obj) in a separate goroutine. This makes obj reachable again, but now without an associated finalizer. Assuming that SetFinalizer is not called again, the next time the garbage collector sees that obj is unreachable, it will free obj.

Reply:
SetFinalizer shouldn't be used for production release. At least there should be an error log to remind developers that they forget to free resources.

Reply:
Let's take this to the extracted PR. I've removed this from here.

Reply:
The finalizer functionallity is not that uncommon to see with these kind of "long-lived" objects: https://github.com/golang/go/blob/d29f5247b8cbf5f2cb7b0e325a5eb1c7c5c1a91f/src/os/file_unix.go#L186